src/app/shared/wysiwyg/prosemirror.markdown.editor.ts

```diff
@@ -17,72 +18,159 @@ export class ProseMirrorImage {
+  update(content: string, images?: ProseMirrorImage[]) {
+    if (content !== this.content || images) {
+      this.content = content;
       this.images = images || this.images;
       let state = EditorState.create(this.stateConfig());
       this.view.update(this.viewProps(state));
     }
   }
 
-  destroy() {
-    this.view.editor.destroy();
+  getContent() {
+    if (this.outputFormat === ProseMirrorFormatTypes.HTML) {
+      return this.removeHTML(this.view.editor.docView.dom.innerHTML);
+    } else {
+      return defaultMarkdownSerializer.serialize(this.view.editor.state.doc);
+    }
   }
 
-  isSelectionEmpty() {
-    return this.view.editor.state.selection.empty;
+  destroy() {
+    this.view.editor.destroy();
   }
 
   viewProps(state: any) {
-    return {
+    let props = {
       state,
-      onAction: (action) => {
-        this.view.updateState(this.view.editor.state.applyAction(action));
-        this.value = defaultMarkdownSerializer.serialize(this.view.editor.state.doc);
-        this.setModel(this.value);
+      dispatchTransaction: (transaction) => {
+        this.view.updateState(this.view.editor.state.apply(transaction));
+        if (this.outputFormat === ProseMirrorFormatTypes.HTML) {
+          let newContent = this.removeHTML(this.view.editor.docView.dom.innerHTML);
+          if (this.content !== newContent) {
+            this.content = newContent;
+            this.setModel(this.content);
+          }
+        } else {
+          this.content = defaultMarkdownSerializer.serialize(this.view.editor.state.doc);
+          this.setModel(this.content);
+        }
       }
     };
+
+    if (!this.editable) {
+      props['editable'] = () => false;
+    }
+
+    return props;
+  }
+
+  isSelectionEmpty() {
+    return this.view.editor.state.selection.empty;
   }
 
   stateConfig() {
-    return {
-      doc: defaultMarkdownParser.parse(this.value ? this.value : ''),
-      plugins: this.buildMenuItemsPlugin()
+    let config = {plugins: this.buildMenuItemsPlugin()};
+    if (this.inputFormat === ProseMirrorFormatTypes.HTML) {
+      // use browser to create element for parsing HTML content
+      let domNode = document.createElement('div');
+      domNode.innerHTML = this.content;
+      config['doc'] = DOMParser.fromSchema(basicSchema).parse(domNode);
+    } else {
+      config['doc'] = defaultMarkdownParser.parse(this.content ? this.content : '');
+    }
+    return config;
+  }
+
+  removeHTML(content) {
+    // prosemirror wraps the single node in a paragraph, when empty prosemirror inserts a <br> into the <p></p>
+    return content.replace('<p>', '').replace('</p>', '').replace('<br>', '');
+  }
+
+  plainTextWithLinks() {
+    let content = '';
+    const getContent = (node: Node) => {
+      if (node.type.name === 'text') {
+        if (content.length > 0 && node.textContent.length > 0
+          && !node.textContent.match(/^\s+/)
+          && !content.match(/^.+\s$/)) {
+          content += ' ';
```

Comment:
regex rookie confirms: this is just inserting a space in case two nodes don't have one between each other, right?
also, i'd vote something like `translatedContent` instead of `content` just to be super clear that that's the bit that's already clean.